### PR TITLE
Add mod "Random" for osu! ruleset

### PR DESCRIPTION
**Pull request by Pasi4K5**
This mod lets the hit objects selected in the settings (including spinners) appear at random positions.
It also works with the Autoplay mod.

Video: https://streamable.com/detzrk

https://user-images.githubusercontent.com/191335/121236934-75806180-c8d1-11eb-85ae-aade018a3008.mov

### CODE REVIEW

**Comment by smoogipoo** [comment]
My take on this is that this is nowhere near usable enough to be a mod, and I'm not sure it can ever be. A lot of work goes into distance snapping and all that, which this is completely throwing away. 
It's nice that you showed the simple section of an easy map, but when any streams hit you have to snap cross screen 1/4 jumps at weird angles, you may as well be playing TAG8.

**Comment by frenzibyte** [comment]
If anything, it could be implemented like in this proposal https://github.com/ppy/osu/issues/9065, offsets the object from its location by a certain distance (not sure how it handles distance snapping though), better than pure randomness.

**Comment by Pasi4K5** [comment]
> If anything, it could be implemented like in this proposal #9065, offsets the object from its location by a certain distance (not sure how it handles distance snapping though), better than pure randomness.

So basically randomise the angle from one hit object to another but keep the distance?

**Comment by peppy** [comment]
keeping distance respecting the original beatmap should work fine, and is the only way i can see this working.

**Comment by Pasi4K5** [comment]
Would you consider something like this more usable?
I set up a short map to test streams and jumps.
**Sliders are not implemented yet.**

This time I tried to keep the distances between the objects as they are in the original beatmap and randomise only the angle.
Also the max. angle is proportional to the distance from the last to the current hit object (like @peppy suggested),
which also prevents too sharp turns in streams.

Video: https://streamable.com/12vvvm

**Comment by peppy** [comment]
Looks like a far better starting point from the video, yep.

**Comment by smoogipoo** [comment]
That last commit can be reverted. SR should change.

**Comment by Pasi4K5** [comment]
> That last commit can be reverted. SR should change.

Since the hit objects are placed randomly to a certain degree, the star rating changes every time the difficulty is calculated though, which also means that the difficulty displayed in the selection differs from the actual one, because the changes are applied when you start the map as well.

Also, how would you calculate the SR of a map that changes every time you play it anyway?

**Comment by smoogipoo** [comment]
I would set a seed in the mod ctor and apply that seed to the RNG (`System.Runtime.Random`). This is actually required for replayability anyway.

From there, actually changing the seed is going to require a bit of work - perhaps it could be exposed as a `Bindable<string>` for the time being (check for usages of `SettingSourceAttribute` in other mods for examples).

**Comment by Pasi4K5** [comment]
That makes sense, but the the star rating would still be different every time the seed is changed.

**Comment by smoogipoo** [comment]
That's fine. Note that it won't work currently because the dynamic diffcalc doesn't take mod settings into account.

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (peppy) [inline osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:32]: Why not make this an int with a large selectable value? I think the UX is better if the user can drag on a slider to change the seed rather than typing a number. Especially since they won't know that it's limited to integer input until they try and fail.

**Comment by peppy** [comment]
I'd also argue that adding a fixed seed makes this mod a lot less fun. I'd rather it changes every play without the user needing to change the seed.

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (Pasi4K5) [reply osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:32]: Wouldn't an `OsuNumberBox` like in the sample implementation here #12602 be more suited for this?

**Comment by Pasi4K5** [comment]
> I'd also argue that adding a fixed seed makes this mod a lot less fun. I'd rather it changes every play/retry without the user needing to change the seed.

What would also be an option is adding a checkbox to the settings. When the checkbox is ticked, a new random seed is generated every time the map is played, if not then you can enter a seed manually. But I'm not sure about this whole changeable seed thing myself honestly.

**Comment by peppy** [comment]
If we want the ability to set a seed then that sounds fine. Having the "refresh seed each retry" as a default selection.

**Comment by smoogipoo** [comment]
@peppy Yeah, I don't disagree. The biggest part is that it needs to be replayable, both for multi and replays. Setting a seed is kinda secondary to that - I just thought it'd be cool if you could list the seed and have others be able to play the map exactly as you did.

Not sure how to make it automatically change right off the bat - perhaps making the seed nullable and having it pick a seed in an `ApplyToX` method.

**Comment by peppy** [comment]
Replays is a good point. Although I'd prefer that in a multiplayer match each player got their own seed, should it not be set by the host.

**Comment by Pasi4K5** [comment]
Mod settings menu looks like this now:



When "Random seed" is enabled (which it is by default) the number box is set to read-only and a random seed is being generated as soon as a beatmap is selected.
I'm not sure if we really need that copy button, but I thought it might be useful.

Atm the seed changes after playing the map as well though, which makes copying the seed from the mod settings menu afterwards impossible. Also it's probably better to change the seed automatically **only** when a beatmap is started, not as soon as another song is selected (as it is the case now). Gonna try to fix that soon.

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:47]: These should not be `static`. See any other mod with settings.

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:60]: What is this for?

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:65]: could just do
```suggestion
            seed.BindValueChanged(e => seed.Default = e.NewValue, true);
```

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:68]: rename the param to `beatmap`

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:53]: Ideally these would be merged into one and be something like

```csharp
[SettingSource("Custom seed", "Custom seed for the random beatmap generation", SettingControlType = typeof(OsuModRandomSettingsControl))]
public Bindable<int?> CustomSeed { get; }
```

Then this whole rigamarole with two properties, setting defaults, confusing naming (i.e. see `ApplyBeatmap()` generating the seed again even though it's already generated in the ctor), etc. goes away entirely.

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:92]: This is redundant. `beatmapHitObject` is already of type `OsuHitObject`.

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:124]: I'd probably bundle these three in a struct type that can be passed around. Would probably clean up `getRandomisedPosition()` too which is quite ugly due to number of params and returning stuff both via return value and `out` parameters.

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:258]: I don't get this. Why is this `finalAngle` calculation not done unconditionally?

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:228]: I don't really understand what this is doing. Do you mind explaining? As far as I can tell this doesn't really "rotate vector towards vector" rather than just perturb the angles between them within a 180 degree interval.

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:273]: `host` should not be static, and the BDL is redundant in its entirety.

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:355]:
```suggestion
                seedNumberBox.Current.Disabled = OsuModRandom.CustomSeedDisabled;
```

looks better. It is very hard to tell by `ReadOnly` that anything has happened at all when the checkbox is toggled.

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:343]: This should not be a thing. The number box already validates characters. At the very most I'd only trim whitespace.

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (Pasi4K5) [reply osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:228]: Let's say `initial` has an angle of 45° and `destination` an angle of 180°.
If `relativeDistance` is set to 0.5, it will return `initial` rotated half way to `destination` (to 112.5°).
If `relativeDistance` is set to 1 it will be rotated all the way to `destination` and if it's 0 nothing will happen.

I made visualisation here: https://www.geogebra.org/calculator/zd7rbe3x
v.initial, v.destination and relativeDistance would be the parameters and v.result would be the returned `Vector2`.
(You can also change relativeDistance and move the points the vectors are pointing to to see how v.result changes.)

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (Pasi4K5) [reply osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:47]: If they aren't `static` the text of the `OsuNumberBox` doesn't change if the value of `seed` is changed and I can't access `random_seed.Value` from `SeedControl.Update()` (which I do with `CustomSeedDisabled` atm (line 55)), which I need to enable/disable the `OsuNumberBox`.

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (Pasi4K5) [reply osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:53]: Not sure if they should be. The reason I separated them is so that you can also see the randomly generated seeds.
And the seed is only generated once in the ctor. That's also what the lines 59 to 60 are doing (see comment below).

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (Pasi4K5) [reply osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:60]: `seed.Default` is -1 by default. If it's still -1 when the ctor is called, `seed.Value` and `seed.Default` are assigned a random seed. So if they already were assigned a random value, it just `return`s.

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (bdach) [reply osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:47]: The proper way to address that is to pass the mod instance down to the control, if absolutely necessary. Static properties are a huge no-no since they will be shared across *all* instances of this class, and you can end up polluting other instances this way.

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (bdach) [reply osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:60]: Yes, I can read. What I was asking about is *when* this can be the case. And this can happen because of all that brokennes of having static setting bindables shared across absolutely all instances of this mod. All this needs to be refactored to not need any statics/shared global state.

File: osu.Game/Screens/Ranking/Expanded/ExpandedPanelMiddleContent.cs
Comment (peppy) [inline osu.Game/Screens/Ranking/Expanded/ExpandedPanelMiddleContent.cs:232]: I'm not sure what this is but it cannot exist.

File: osu.Game/Screens/Ranking/ScorePanel.cs
Comment (peppy) [inline osu.Game/Screens/Ranking/ScorePanel.cs:210]: Same goes for this. Your PR should not touch anything outside of your mod classes. Anything else that needs changing should be its own PR which has zero dependencies on your mod.

Comment (Pasi4K5) [reply osu.Game/Screens/Ranking/ScorePanel.cs:210]: Alright, but before I submit another PR: Should the seed even be displayed at the result screen? I don't want to implement something that makes it possible for mods to modify the result screen when it ends up never being used, because this mod is probably the only use case for something like this.

Comment (bdach) [reply osu.Game/Screens/Ranking/ScorePanel.cs:210]: Mod icons have tooltips that are supposed to be showing the mod customisation options set for a given score.

Leaderboards for this sort of mod are pretty much useless in the current state of things anyway. And making them not useless for random is a whole lot of work, outside the scope of this PR.

Comment (Pasi4K5) [reply osu.Game/Screens/Ranking/ScorePanel.cs:210]: I didn't know about the tooltips to be honest. That makes this redundant of course.

**Comment by peppy** [comment]
I started tiding this up only to realise that it's actually unplayable due to sliders seemingly not being shifted correctly. Is that intended? It causes incorrect jumps throughout any beatmap with sliders present.

**Comment by Pasi4K5** [comment]
I didn't implement slider randomisation yet (there's also a TODO for that at line 92). This is gonna get quite tricky because I have to find a way to keep the distance to the sliders while also making sure that the sliders stay entirely inside of the playfield, which is a problem with big sliders especially.

**Comment by Pasi4K5** [comment]
Slider positions are now being randomised as well.
Mod should be in a playable state now.

(Actually sometimes it even lowers the star diff slightly.)

When sliders are either really big or really close to the border of the playfield, the distances might still be slightly off.

Video: https://streamable.com/oyyn0n

**Comment by Susko3** [comment]
Slider ticks appear in the wrong location. Can be visible on longer sliders when using osu!lazer skin.

Normal:

osu_2021-05-15_01-26-45

Random:

osu_2021-05-15_01-27-16

**Comment by peppy** [comment]
@Pasi4K5 I've tidied this up and would like to get it in as an initial implementation, without worrying too much about the algorithm for now (that can be improved going forward).

There's one change which I think needs to happen before merging though: The first hitobject in the beatmap should get a 100% random position. This should also happen for the first object after a spinner.

This can be easily achieved by making `prevObjectInfo` a `RandomisedObjectInfo?` and nulling it initially (removing the `i==0` case i added) and on reaching a spinner. You will need to make the randomisation method support this though, as currently it is always working with a previous value.

Let me know if you aren't confident in making the change and I'll look into it further; felt it's best to leave it to you if you're willing to fix it up.

**Review by peppy — CHANGES_REQUESTED** [review]
As proposed

**Comment by Pasi4K5** [comment]
Actually I just noticed something else:

`moveSliderIntoPlayfield()` only moves the slider's control points into the playfield, but that doesn't necessarily mean that the whole slider is inside of it, so parts of sliders are off screen sometimes. Fixing it should't be a big problem though. I'll just make sure that all the slider ticks are inside of the playfield too.

**Comment by Pasi4K5** [comment]
Sliders should always be entirely inside of the playfield now (except if they are bigger than the playfield obviously).

I don't know if `MarginPadding` should be used for this, because that's probably not what it's meant for, but the property names are really fitting and I thought it was unnecessary to create a new `struct` just for this.
https://github.com/ppy/osu/blob/bdbd64c88d6a9cace3862fcdd45ccb6109d38cda/osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs#L165

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (peppy) [inline osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:187]: This should probably be moved to the same method as the other call to the same function. It's hard to follow if it's done in two separate places.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs
Comment (peppy) [reply osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs]: I've moved this for you, but it needs an inline comment explaining why these two shifts are required in the first place.

**Comment by Pasi4K5** [comment]
After playing a few maps with RD I got a few ideas on how to improve the randomisation.
Streams work very well with this mod, but when converting jumps it could be way better. (They are more like really far spaced streams rather than jumps in many cases, especially with short jumps.)
Should I just leave it like that for now and open an issue for tracking improvements (including the problem i mentioned earlier https://github.com/ppy/osu/pull/12568#issuecomment-841015003) as soon as it gets merged?

**Comment by peppy** [comment]
unless the changes are going to simplify the code, i would hold off until merge.

**Comment by peppy** [comment]
While this is by no means perfect, I think it's a great start and getting it in the hands of more users will lead to further refactoring/changes (if it gains traction).

**Review by peppy — APPROVED** [review]

**Pull request merged**
